reporting/impl/src/main/java/org/jboss/windup/reporting/export/ExportCSVFileRuleProvider.java

```diff
@@ -183,50 +188,140 @@ private String buildLinkString(Iterable<LinkModel> links)
+                        "File Path", "Line", "Story points", "Parent Application"};
                 String mergedFilename = PathUtil.cleanFileName(MERGED_CSV_FILENAME) + ".csv";
-                CSVWriter mergedFileWriter = initCSVWriter(outputFolderPath + mergedFilename, true);
+                CSVWriter mergedFileWriter = initCSVWriter(outputFolderPath + mergedFilename, headerLine);
                 projectToFile.put(MERGED_CSV_FILENAME, mergedFileWriter);
             }
             if (!projectToFile.containsKey(projectModel.getName()))
             {
+                String[] headerLine;
+                headerLine = new String[]{"Rule Id", "Issue Category", "Title", "Description", "Links", "Application", "File Name",
+                        "File Path", "Line", "Story points"};
                 String filename = PathUtil.cleanFileName(projectModel.getRootFileModel().getFileName()) + ".csv";
-                CSVWriter writer = initCSVWriter(outputFolderPath + filename, false);
+                CSVWriter writer = initCSVWriter(outputFolderPath + filename, headerLine);
                 projectToFile.put(projectModel.getName(), writer);
                 LOG.info("Setting csv filename to: " + filename + " for id: " + projectModel.getId());
                 projectModel.setCsvFilename(filename);
             }
-            projectToFile.get(projectModel.getName()).writeNext(line);
-            //Convert line array to ArrayList, add extra field for merged file on the end,
-            // then convert back to array to send to CSVWriter
-            ArrayList<String> mergedList = new ArrayList<String>(Arrays.stream(line).collect(Collectors.toList()));
-            mergedList.add(projectModel.getRootFileModel().asFile().getName());
-            String[] mergedLine = new String[ mergedList.size() ];
-            projectToFile.get(MERGED_CSV_FILENAME).writeNext(mergedList.toArray(mergedLine));
+            if (!projectToFile.containsKey(APP_FILE_TECH_CSV_FILENAME) && isLineForAppTagFile)
+            {
+                String[] headerLine;
+                headerLine = dynamicHeaderLine;
+                String appFileTechFilename = PathUtil.cleanFileName(APP_FILE_TECH_CSV_FILENAME) + ".csv";
+                CSVWriter appFileTechWriter = initCSVWriter(outputFolderPath + appFileTechFilename, headerLine);
+                projectToFile.put(APP_FILE_TECH_CSV_FILENAME, appFileTechWriter);
+            }
+            if (isLineForAppTagFile)
+            {
+                projectToFile.get(APP_FILE_TECH_CSV_FILENAME).writeNext(line);
+            }
+            else
+            {
+                projectToFile.get(projectModel.getName()).writeNext(line);
+                //Convert line array to ArrayList, add extra field for merged file on the end,
+                // then convert back to array to send to CSVWriter
+                ArrayList<String> mergedList = new ArrayList<String>(Arrays.stream(line).collect(Collectors.toList()));
```

Comment:
is not Collectors.toList() returning an ArrayList<String> ?

Reply:
Removing the cast causes this error: 

> Incompatible types. Required ArrayList<String> but 'collect' was inferred to R: no instance(s) of type variable(s) T exist so that List<T> conforms to ArrayList<String>

>

Reply:
I see.
But I think that using the interface would be eough : 
List<String> mergedList = Arrays.stream(line).collect(Collectors.toList());
Or specifying to type directly on the stream:
ArrayList<String> mergedList = Arrays.stream(line).collect(toCollection(ArrayList::new));

But I know it's only a matter of taste.... doing the new ArrayList doesn't do any harm.